tests/templates/test_integration.py

```diff
@@ -176,558 +138,331 @@
-    """CV qnode factory passing differentiable parameters as positional arguments"""
-
-    # Signature juggling to pass diffable as single positional arg, but keep track of input names
-
-    @qml.qnode(dev, interface=interface)
-    def circuit(*diffable, keys_diffable1=None, keys_diffable2=None, nondiffable1=None, nondiffable2=None):
-        # Separate differentiable arguments
-        diffable1 = diffable[:n_args1]
-        diffable2 = diffable[n_args1:]
-
-        # Turn diffables back into dictionaries
-        dict1 = {key: item for key, item in zip(keys_diffable1, diffable1)}
-        dict2 = {key: item for key, item in zip(keys_diffable2, diffable2)}
-
-        # Merge with nondiffables
-        dict1.update(nondiffable1)
-        dict2.update(nondiffable2)
-
-        # Add number of wires
-        dict1['wires'] = range(2)
-        dict2['wires'] = range(2)
-
-        # Actual circuit
-        qml.Displacement(1., 1., wires=0)
-        template1(**dict1)
-        template2(**dict2)
-        qml.Displacement(1., 1., wires=1)
-        return [qml.expval(qml.Identity(0)), qml.expval(qml.X(1))]
-
-    return circuit
-
-
-def qnode_qubit_kwargs(dev, interface, template1, template2):
-    """Qubit qnode factory passing differentiable parameters as keyword arguments"""
-
-    @qml.qnode(dev, interface=interface)
-    def circuit(nondiffable1=None, nondiffable2=None):
-        # Add wires
-        nondiffable1['wires'] = range(2)
-        nondiffable2['wires'] = range(2)
-
-        # Circuit
-        qml.PauliX(wires=0)
-        template1(**nondiffable1)
-        template2(**nondiffable2)
-        qml.PauliX(wires=1)
-        return [qml.expval(qml.Identity(0)), qml.expval(qml.PauliX(1))]
-
-    return circuit
-
-
-def qnode_cv_kwargs(dev, interface, template1, template2):
-    """CV qnode factory passing differentiable parameters as keyword arguments"""
-
-    @qml.qnode(dev, interface=interface)
-    def circuit(nondiffable1=None, nondiffable2=None):
-        # Add wires
-        nondiffable1['wires'] = range(2)
-        nondiffable2['wires'] = range(2)
-
-        # Circuit
-        qml.Displacement(1., 1., wires=0)
-        template1(**nondiffable1)
-        template2(**nondiffable2)
-        qml.Displacement(1., 1., wires=1)
-        return [qml.expval(qml.Identity(0)), qml.expval(qml.X(1))]
-
-    return circuit
-
-
-######################
-
-
-class TestIntegrationCircuit:
-    """Tests the integration of templates into circuits using different interfaces. """
-
-    @pytest.mark.parametrize("template1, diffable1, nondiffable1", QUBIT_DIFFABLE_NONDIFFABLE)
-    @pytest.mark.parametrize("template2, diffable2, nondiffable2", QUBIT_DIFFABLE_NONDIFFABLE)
+# List templates in NO_OP_BEFORE that do not allow for operations
+# before they are called in a quantum function.
+# These templates will be skipped in tests of that nature.
+
+NO_OP_BEFORE = [qml.templates.AmplitudeEmbedding, qml.templates.BasisEmbedding]
```

Comment:
Out of curiosity: shouldn't state preparations (e.g. BasisState and Mottonen) be included here? Also curious about the other types of embeddings.

Reply:
Very good question!

Reply:
That was quite involved:

- ``MottonenStatePrep`` can be used with preceding operations
- ``BasisState`` is not a template but an ``op``, so not explicitly tested here
- ``BasisStatePreparation`` *is* a template and tested, but implemented by PauliX rotations and can therefore be used with preceding operations
- ``AmplitudeEmbedding`` uses the op ``QubitStateVec`` on ``default.qubit`` and cannot be used with preceding operations
- ``BasisEmbedding`` was using the op ``BasisState``, but I changed it to be implemented with PauliX rotations, so this edge case disappears

Your comment uncovered a bug in the tests - using the type checking works for ``ops`` but not for ``templates``, since they are functions. So checking for ``NO_OP_BEFORE`` accidentally excluded *all* templates from that test case. This is fixed now by instead checking the name of the function.

Reply:
Oh, that's really nice, thanks for the details! :blush: :1st_place_medal:

Reply:
(A follow-up for a final clarification): these descriptions are about how each can be used inside of a `qfunc` functionally. Would it still make sense to (at one point, definitely out of scope here) add a custom error to each of them if they are not the first operation in the `qfunc`?

Reply:
Definitely. And here it will be useful to have a template class, see also #574, #575...